### ISSUE TITLE
Evaluate PDFMiner, pdf-extractor, pdfreader, and ghostscript with a possibility of including them as possible extractors

### ISSUE THREAD

**Issue by aarohijohal** [axa-group/Parsr#3]
Recent tests have shown that PDFMiner outperforms pdf2json as a pdf extractor. It's pdf2txt.py -A -t xml -o <output.xml> <input.pdf> provides an output xml very similar in content to the output of pdf2json.
It could be nice to provide PDFMiner as an extaction option.

Note: PDFMiner's installation depends on python2 and not python3.

Other extractors to be evaluated:

- https://www.npmjs.com/package/pdfreader
- https://www.npmjs.com/package/pdf-extractor
- Ghostscript: see https://stackoverflow.com/a/6189489

**Comment by aarohijohal** [comment]
## pdfminer
- ➕ bounding boxes by char, word
- ➕ text seems to be clearly interpreted
- ➖ depends on python2
- ➖ to get a proper divison of words, the proper values for 
  ```
  -M char_margin
  -L line_margin
  -W word_margin 
  ```
  have to be specified. The default value seems to be alright, but we need to make sure we use the best set of defaults using this page for reference: https://euske.github.io/pdfminer/

Example XML output:
```<?xml version="1.0" encoding="utf-8" ?>
<pages>
<page id="1" bbox="0.000,0.000,612.000,792.000" rotate="0">
<textbox id="0" bbox="96.910,578.887,464.695,602.762">
<textline bbox="96.910,578.887,464.695,602.762">
<text font="MNRSXS+LinLibertine" bbox="96.910,578.887,109.029,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">D</text>
<text font="MNRSXS+LinLibertine" bbox="109.029,578.887,116.757,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">é</text>
<text font="MNRSXS+LinLibertine" bbox="116.636,578.887,125.228,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">v</text>
<text font="MNRSXS+LinLibertine" bbox="125.107,578.887,132.834,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">e</text>
<text font="MNRSXS+LinLibertine" bbox="132.834,578.887,137.398,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">l</text>
<text font="MNRSXS+LinLibertine" bbox="137.398,578.887,146.112,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">o</text>
<text font="MNRSXS+LinLibertine" bbox="146.112,578.887,155.084,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">p</text>
<text font="MNRSXS+LinLibertine" bbox="155.084,578.887,164.057,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">p</text>
<text font="MNRSXS+LinLibertine" bbox="164.143,578.887,171.871,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">e</text>
<text font="MNRSXS+LinLibertine" bbox="171.871,578.887,185.528,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">m</text>
<text font="MNRSXS+LinLibertine" bbox="185.528,578.887,193.256,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">e</text>
<text font="MNRSXS+LinLibertine" bbox="193.256,578.887,202.626,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">n</text>
<text font="MNRSXS+LinLibertine" bbox="202.626,578.887,208.089,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">t</text>
<text> </text>
<text font="MNRSXS+LinLibertine" bbox="212.307,578.887,221.055,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">d</text>
<text font="MNRSXS+LinLibertine" bbox="221.055,578.887,225.688,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">’</text>
<text font="MNRSXS+LinLibertine" bbox="225.688,578.887,234.868,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">u</text>
<text font="MNRSXS+LinLibertine" bbox="234.868,578.887,244.238,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">n</text>
<text font="MNRSXS+LinLibertine" bbox="244.238,578.887,251.966,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">e</text>
<text> </text>
<text font="MNRSXS+LinLibertine" bbox="256.184,578.887,264.085,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">a</text>
<text font="MNRSXS+LinLibertine" bbox="264.085,578.887,273.057,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">p</text>
<text font="MNRSXS+LinLibertine" bbox="273.057,578.887,282.030,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">p</text>
<text font="MNRSXS+LinLibertine" bbox="282.030,578.887,286.594,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">l</text>
<text font="MNRSXS+LinLibertine" bbox="286.594,578.887,291.279,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">i</text>
<text font="MNRSXS+LinLibertine" bbox="291.279,578.887,298.678,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">c</text>
<text font="MNRSXS+LinLibertine" bbox="298.678,578.887,306.579,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">a</text>
<text font="MNRSXS+LinLibertine" bbox="306.579,578.887,312.042,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">t</text>
<text font="MNRSXS+LinLibertine" bbox="312.042,578.887,316.727,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">i</text>
<text font="MNRSXS+LinLibertine" bbox="316.727,578.887,325.440,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">o</text>
<text font="MNRSXS+LinLibertine" bbox="325.440,578.887,334.810,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">n</text>
<text> </text>
<text font="MNRSXS+LinLibertine" bbox="339.028,578.887,347.776,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">d</text>
<text font="MNRSXS+LinLibertine" bbox="347.776,578.887,355.504,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">e</text>
<text> </text>
<text font="MNRSXS+LinLibertine" bbox="359.705,578.887,368.453,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">d</text>
<text font="MNRSXS+LinLibertine" bbox="368.453,578.887,373.138,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">i</text>
<text font="MNRSXS+LinLibertine" bbox="373.138,578.887,383.199,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">ﬀ</text>
<text font="MNRSXS+LinLibertine" bbox="383.199,578.887,392.379,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">u</text>
<text font="MNRSXS+LinLibertine" bbox="392.379,578.887,399.122,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">s</text>
<text font="MNRSXS+LinLibertine" bbox="399.122,578.887,403.807,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">i</text>
<text font="MNRSXS+LinLibertine" bbox="403.807,578.887,412.520,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">o</text>
<text font="MNRSXS+LinLibertine" bbox="412.520,578.887,421.890,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">n</text>
<text> </text>
<text font="MNRSXS+LinLibertine" bbox="426.108,578.887,434.700,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">v</text>
<text font="MNRSXS+LinLibertine" bbox="434.700,578.887,439.385,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">i</text>
<text font="MNRSXS+LinLibertine" bbox="439.385,578.887,448.133,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">d</text>
<text font="MNRSXS+LinLibertine" bbox="448.133,578.887,455.861,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">é</text>
<text font="MNRSXS+LinLibertine" bbox="455.982,578.887,464.695,602.762" colourspace="DeviceGray" ncolour="None" size="23.875">o</text>
<text>
</text>
</textline>
</page>
</pages>
```

## pdfreader
- ➖ depends on pdf2json entirely (https://github.com/modesty/pdf2json), thus as good as that extractor.
- ➖ a simple table reconstruction based on matrix/pattern detection (arrangement of cols, rows) on top of the pdf2json output: https://github.com/adrienjoly/npm-pdfreader/blob/master/lib/TableParser.js
- ➖ table detection doesn't seem to have any delta - coordinates for the content have to be perfectly aligned for cells to be properly picke up.
- ➕ features a regex matcher called 'sequential fields' here: https://github.com/adrienjoly/npm-pdfreader/blob/master/lib/SequentialParser.js

## pdf-extractor
- ➖ a wrapper around pdf.js to generate html output for compatibility with Box View.
Hence, see the following option:

## pdf-js (and pdf-js.extract? https://www.npmjs.com/package/pdf.js-extract)
- ➕ Mozilla's extractor can also be used to generate an output of this sort:

```
{
    "filename": "helloworld.pdf",
    "meta": {
        "info": {
            "PDFFormatVersion": "1.7",
            "IsAcroFormPresent": false,
            "IsXFAPresent": false
        },
        "metadata": null
    },
    "pages": [
        {
            "pageInfo": {
                "num": 1,
                "scale": 1,
                "rotation": 0,
                "offsetX": 0,
                "offsetY": 0,
                "width": 200,
                "height": 200
            },
            "content": [
                {
                    "x": 70,
                    "y": 150,
                    "str": "Hello, world!",
                    "dir": "ltr",
                    "width": 64.656,
                    "height": 12,
                    "fontName": "Times"
                }
            ]
        }
    ],
    "pdfInfo": {
        "numPages": 1,
        "fingerprint": "1ee9219eb9eaa49acbfc20155ac359c3"
    }
}
```

**Comment by aarohijohal** [comment]
After careful assessement, `pdfminer` seems to solve the need well. Closed after merge of the PR #32
